modules/007-registry-packages-proxy/images/registry-packages-proxy/src/internal/credentials/watcher.go

```diff
@@ -0,0 +1,216 @@
+	"time"
+
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	toolsWatch "k8s.io/client-go/tools/watch"
+
+	"github.com/deckhouse/deckhouse/go_lib/registry-packages-proxy/registry"
+)
+
+type Watcher struct {
+	k8sClient                     *kubernetes.Clientset
+	k8sDynamicClient              dynamic.Interface
+	registrySecretDiscoveryPeriod time.Duration
+	registryClientConfigsMutex    sync.RWMutex
+	registryClientConfigs         map[string]*registry.ClientConfig
+	logger                        *log.Entry
+}
+
+func NewWatcher(k8sClient *kubernetes.Clientset, k8sDynamicClient dynamic.Interface, registrySecretDiscoveryPeriod time.Duration, logger *log.Entry) *Watcher {
+	return &Watcher{
+		k8sClient:                     k8sClient,
+		k8sDynamicClient:              k8sDynamicClient,
+		registrySecretDiscoveryPeriod: registrySecretDiscoveryPeriod,
+		registryClientConfigs:         make(map[string]*registry.ClientConfig),
+		logger:                        logger,
+	}
+}
+
+func (w *Watcher) Get(repository string) (*registry.ClientConfig, error) {
+	w.registryClientConfigsMutex.RLock()
+	defer w.registryClientConfigsMutex.RUnlock()
+
+	clientConfig, ok := w.registryClientConfigs[repository]
+	if !ok {
+		return nil, errors.Errorf("registry client config for repository '%s' not found", repository)
+	}
+
+	return clientConfig, nil
+}
+
+func (w *Watcher) Watch(ctx context.Context) {
+	var wg sync.WaitGroup
+
+	wg.Add(2)
+
+	go func() {
+		defer wg.Done()
+
+		w.watchSecret(ctx)
+	}()
+
+	go func() {
+		defer wg.Done()
+
+		w.watchModuleSources(ctx)
+	}()
+
+	wg.Wait()
+}
+
+func (w *Watcher) watchSecret(ctx context.Context) {
+	err := w.fetchSecret(ctx)
+	if err != nil {
+		w.logger.Errorf("Fetch secret: %v", err)
+		return
+	}
+
+	ticker := time.NewTicker(w.registrySecretDiscoveryPeriod)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			err := w.fetchSecret(ctx)
+			if err != nil {
+				w.logger.Errorf("Fetch secret: %v", err)
+				return
+			}
+		}
+	}
+}
+
+func (w *Watcher) fetchSecret(ctx context.Context) error {
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+
+	// Get the secret with the registry credentials
+	secret, err := w.k8sClient.CoreV1().Secrets("d8-system").Get(ctx, "deckhouse-registry", metav1.GetOptions{})
+	if err != nil {
+		return errors.Wrap(err, "failed to get secret")
+	}
+
+	var input registrySecretData
+
+	input.FromSecretData(secret.Data)
+
+	registryConfig, err := input.toClientConfig()
+	if err != nil {
+		return errors.Wrap(err, "failed to convert secret data to registry config")
+	}
+
+	w.registryClientConfigsMutex.Lock()
+	w.registryClientConfigs[""] = registryConfig
```

Comment:
please use constant with comment for empty string case

Reply:
done